[clang] Fix self-capturing `__block` variables

ille-apple opened the pull request:
This is an updated version of https://reviews.llvm.org/D90434 from 2020.  Due to time constraints I stopped working on the patch, and unfortunately never got back around to it until now.

Clang special-cases `__block` variables whose initializer contains a block literal that captures the variable itself.  But this special case doesn't work for variables of record types, causing Clang to emit broken code.

Fix this by initializing such variables directly on the heap, and add a warning for when this needs to be done.

### Background on the special case

`__block` variables are unique in that their address can change during the lifetime of the variable.  Each `__block` variable starts on the stack.  But when a block is retained using `Block_copy`, both the block itself and any `__block` variables it captures are moved to the heap.

What's tricky is when this move happens while the variable is still being initialized.  For example:

```c
int copyBlockAndReturnInt(void (^block)(void)) {
    Block_copy(block);
    return 42;
}

int main() {
  __block int val = copyBlockAndReturnInt(^{
      printf("%d\n", val);
  });
}
```

Here, when `Block_copy` is called, `val` is moved to the heap while still in an uninitialized state.  Then when `copyBlockAndReturnInt` returns, `main` needs to store the return value to the variable's new location, not to the original stack slot.

This can only happen if a block literal that captures the variable is located syntactically within the initializer itself.  If it's before the initializer then it couldn't capture the variable, and if it's after the initializer then it couldn't execute during initialization.

So Clang detects when such a literal exists inside the initializer; `CGDecl.cpp` refers to this condition as `capturedByInit`.  If it does, then the initialization code operates in a special mode.  In several functions in that file, there's an `LValue` object and an accompanying `capturedByInit` flag. If `capturedByInit` is false (the typical case), then the lvalue points to the object being initialized.  But if it's true, the lvalue points to the byref header.  Then, depending on the type of the variable, you end up at one of many checks of the form:

```cpp
if (capturedByInit)
  drillIntoBlockVariable(*this, lvalue, cast<VarDecl>(D));
```

`drillIntoBlockVariable` emits a load of the forwarding pointer from the byref header (to find the object's current location), and then modifies `lvalue` to point to the object, unifying the two cases.  Importantly, this happens at the last minute: after emitting code to evaluate the initializer, but before emitting the `store` instruction to write the evaluated value into the variable.

Okay, but then why are there two different code paths?  If the address calculation can be safely deferred, why not just do that all the time rather than having a separate mode just for self-capturing `__block` variables?

Because it *can't* always be safely deferred.

### Harder cases

First, consider the case where variable being initialized is a C struct.  This is already problematic.  `drillIntoBlockVariable` is supposed to be called after evaluating the initializer but before doing a `store`.  But when initializing a struct, Clang doesn't evaluate the whole initializer before storing.  Instead it evaluates the first field, stores the first field, evaluates the second field, stores the second field, and so on.  This is actually necessary for correctness[^1] given code like:

```c
struct Foo { int a, b; };
struct Foo foo = {1, foo.a};
```

But if any one of the field evaluations can move the object, then Clang would need to reload the forwarding pointer before storing each field.  And that would need to apply recursively to any subfields.  Such a scheme is definitely possible, but Clang doesn't implement it.

And if instead of a C struct you have a C++ class, supporting relocation to the heap becomes flat-out impossible.  Consider:

```cpp
struct Foo {
    void (^block_)(void);
    Foo(void (^block)(void)) {
        printf("I am at %p\n", this);
        block_ = Block_copy(block);
        printf("I am now at %p\n", this); // shouldn't change!
    }
};

__block Foo foo(^{
    printf("%p\n", &foo);
});
```

C++ constructors can observe the address of the object being constructed, and expect that address not to change.  Once the object has finished being constructed, it can be relocated by calling the move or copy constructor, and that's what Clang uses for block captures in the non-self-capturing case.  But in this case, `Block_copy` is called while the stack object is still running its constructor.  It would be nonsensical to try to move out of it in this state.

### Clang's current behavior

So what does Clang currently do?  Well, for both C structs and C++ classes (as well as unions[^2]), it runs into this TODO dating to 2011:

```cpp
case TEK_Aggregate:
  [...]
    // TODO: how can we delay here if D is captured by its initializer?
```

Clang simply neglects to call `drillIntoBlockVariable` and falls into its normal initialization code.  That means the lvalue is initialized as if it pointed to the object, when it actually points to the byref header.  As such, the generated code is non-functional.  And this is true for all `__block` variables of record type that Clang detects as self-capturing, even if the initializer never actually calls `Block_copy`.

### Solution

As stated, it's impossible for self-capturing variables of at least some record types to support moving to the heap.  But there is a way to make them work regardless.  Any given `__block` variable can move at most once, from the stack to the heap.  Once on the heap, it never needs to move again.  So, as suggested by @rjmccall in 2020, we can create a heap allocation before initialization even starts, and then initialize the object directly on the heap, knowing its address will remain stable.

This patch applies this approach to all self-capturing `__block` variables with record type, i.e. all variables that previously triggered the broken code generation.

The obvious downside of the approach that we pay the cost of a heap allocation even if `Block_copy` is never called.  Conveniently, this can never be a performance regression *per se*, since the affected cases didn't work at all before.  Still, there is room for future improvement, such as by making mid-initialization relocation work for C structs[^3], or detecting initializers that do contain a block literal but can't possibly call `Block_copy` on it.

To assist performance-sensitive codebases, this patch adds a warning, `-Wimplicit-block-var-alloc`, that triggers whenever a block requires an implicit allocation.  Since most codebases probably don't care, this warning is disabled by default and is not included in any standard warning groups.

### Allocation failure

Another downside of the approach is that because the heap allocation is completely implicit, there is no way to check for allocation failure.  Normally, heap allocation happens only when you call `Block_copy`, which means you can catch allocation failure by checking whether `Block_copy` returns NULL.

...Or so I thought.  It turns out that checking `Block_copy`'s return value is only a half-measure.  `Block_copy` does return NULL if there's an allocation failure when moving the block itself to the heap.  But then `Block_copy` calls into the copy helper to move captured variables to the heap, which can also fail.  If it does, the copy helper has no way to report that failure.  Instead you just get a crash (which should probably be a proper abort; Apple folks, see rdar://126632757).  This limitation is more or less baked into the ABI, since copy helpers return void.  Even xnu, which uses blocks in kernel space, does not support allocation failure for them - though, luckily, xnu's version [waits for memory to be available rather than crashing](https://github.com/apple-oss-distributions/xnu/blob/94d3b452840153a99b38a3a9659680b2a006908e/libkern/libclosure/runtime.cpp#L49).

The point is, if recovery from allocation failure is already not a thing, then this patch isn't responsible for breaking it.  Even if blocks are changed someday to support it, it would probably be desired only by a tiny minority of codebases, and those codebases could avoid implicit allocations by enabling the new warning and setting it as an error.

### ABI limitation

The blocks runtime has no API to allocate a byref struct directly on the heap, because it's never been necessary before.  There is only `_Block_object_assign` to move one from the stack to the heap.  It would be nice to add an API to allocate directly on the heap, but for the dual sakes of backward compatibility and laziness, I have not attempted to do so.

Instead, the codegen works with the existing runtime, as follows.  First, a byref struct is allocated on the stack (including space for the object), but only the header is initialized; the object part is left uninitialized.  Then `_Block_object_assign` is called to migrate it to the heap (which unfortunately pays the cost of memcpying the uninitialized stack memory to the heap).  This produces the desired heap allocation, but also increases the reference count, which is balanced by immediately calling `_Block_object_dispose`.  After that, the object can finally be initialized inside the new heap allocation.

### Improved self-capture analysis

The existing check for self-capturing `__block` variables is implemented in the function `isCapturedBy`.  But it turns out that that implementation is overcomplicated for no good reason.  In ancient days (2011), `isCapturedBy` only worked on `Expr`s, and it recursed on the `Expr`'s children, which it assumed would also be `Expr`s.  This assumption turned out to be false for GNU statement expressions.  The simple fix would have been to change `isCapturedBy` to accept and recurse on arbitrary `Stmt`s instead, since `children()` is defined for all `Stmt`s, not just `Expr`s.  I'm not sure if there was some reason that wasn't possible at the time, but what happened instead is that a special case was added for statement expressions, which over a series of patches got more and more convoluted.

This patch replaces the whole thing with an `EvaluatedExprVisitor` (another suggestion from @rjmccall).  This is not only simpler, but also more precise, eliminating false positive self-captures including:

- Block literals in unevaluated contexts (`sizeof(^{ ... })`).

- Some types of statements inside GNU statement expressions.

This patch also moves the self-capture check earlier in the compilation process, from CodeGen to Sema.  This is necessary in order to support the aforementioned warning.

### Constant initialization

This patch also fixes a related bug where self-capturing `__block` variables could trigger this assert:

```cpp
assert(!capturedByInit && "constant init contains a capturing block?");
```

The assert is important, because constant initialization doesn't support the `drillIntoBlockVariable` mechanism.  Now, a constant initializer obviously can't include a call to `Block_copy`; in fact, it can't include block literals at all unless they have no captures.  But I found two cases where the assert can be triggered anyway.

First, the initializer can syntactically contain a capturing block literal but never evaluate it:

```cpp
__block constexpr int foo = true ? 1 : ^{ return foo; }();
```

This is addressed in this patch by forcing constant-initialized variables to be treated as non-self-capturing.

Second, the initializer can simply not be constant.  With variables of types like `const int`, Clang first tries to constant-evaluate the initializer, then falls back to runtime initialization if that fails.  But the assert is located in a place where it triggers just based on the try.  This is addressed in this patch by moving the assert down and conditioning it on constant evaluation succeeding.

[^1]: Supporting such code appears to be required by the C++ spec (see
    https://timsong-cpp.github.io/cppwp/n4861/dcl.init.aggr#6), though not
    by the C spec (see https://port70.net/~nsz/c/c11/n1570.html#6.7.9p23).

[^2]: There are other `TEK_Aggregate` types besides records, but all such
    types either can't be captured by blocks (arrays) or can't be the
    type of a local variable at all.

[^3]: Though even in C, moving mid-initializer could break code that took the address of the variable:

    ```c
    struct Foo { void *a; void *(^b)(void); };
    __block struct Foo x = {&x.b, Block_copy(^{ return x.a; })};
    assert(x.a == &x.b); // fails without up-front heap allocation
    ```

    However, it's a bit less obvious that such code *needs* to work; perhaps the user should just be aware the variable will be moving.  It's not like C++ where construction is clearly an indivisible operation performed at a single address.

ille-apple committed:
[clang] Fix self-capturing `__block` variables

Clang special-cases `__block` variables whose initializer contains a
block literal that captures the variable itself.  But this special case
doesn't work for variables of record types, causing Clang to emit broken
code.

Fix this by initializing such variables directly on the heap, and add a
warning for when this needs to be done.

rdar://70498802

ille-apple commented:
Rebased, because the original version failed CI due to a regression on `main` (#89476).

ille-apple commented:
Ping :)

ille-apple commented:
Rebased due to merge conflict.

ChuanqiXu9 reviewed (commented):
Serialization related change looks trivial and good. But I feel better to leave the formal approval to CG part reviewers.

ille-apple commented:
Weekly ping.

rjmccall commented on clang/lib/CodeGen/CGBlocks.cpp:
```diff
@@ -2499,21 +2501,21 @@ CodeGenFunction::buildByrefHelpers(llvm::StructType &byrefType,
-      // Otherwise, we transfer ownership of the retain from the stack
-      // to the heap.
+        // Otherwise, we transfer ownership of the retain from the stack
+        // to the heap.
```
The indentation here is correct — it's a continuation of the same thought as the comment before the `if`.

ille-apple replied:
Makes sense, but the re-indentation came from clang-format (because I modified the code next to it), and not doing it will upset the CI checks.

How about I move the first comment after the `if` and indent both of them?
```cpp
      if (type->isBlockPointerType()) {
        // Block pointers need to be copied, and there's no direct
        // transfer possible.
        return ::buildByrefHelpers(CGM, byrefInfo,
                                   ARCStrongBlockByrefHelpers(byrefInfo));
      } else {
        // Otherwise, we transfer ownership of the retain from the stack
        // to the heap.
        return ::buildByrefHelpers(CGM, byrefInfo,
                                   ARCStrongByrefHelpers(byrefInfo));
      }
```

rjmccall commented on clang/lib/CodeGen/CGBlocks.cpp:
```diff
@@ -2653,11 +2655,39 @@ const BlockByrefInfo &CodeGenFunction::getBlockByrefInfo(const VarDecl *D) {
+  // the stack (with the structural components initialized but not the object
+  // itself), then call _Block_object_assign to move it to the heap (which is
+  // safe because we forced a no-op copy helper), then call
+  // _Block_object_dispose to release the extra ref from _Block_object_assign.
```
Ouch.  Yeah, it'd be great if we could get a new runtime function to optimize this, but you're right, this is what we have to do right now.

rjmccall commented on clang/lib/CodeGen/CGBlocks.cpp:
```diff
@@ -2653,11 +2655,39 @@ const BlockByrefInfo &CodeGenFunction::getBlockByrefInfo(const VarDecl *D) {
+  RawAddress out = CreateDefaultAlignTempAlloca(P->getType(), "initOnHeap.out");
+
+  llvm::Value *args[] = {Builder.CreateBitCast(out.getPointer(), VoidPtrTy),
+                         Builder.CreateBitCast(P, VoidPtrTy),
```
These bitcasts should no longer be necessary, right?  Do we have a mode that still generates typed pointers in IR?

If you're trying to handle address-space differences, this is the wrong way to do it.

ille-apple replied:
Ah, this is an artifact of the change having originally been written in 2020.  Looks like I can just drop the bitcast.  I don't think address-space differences are possible.

rjmccall commented on clang/lib/CodeGen/CGDecl.cpp:
```diff
@@ -2120,27 +2066,33 @@ void CodeGenFunction::EmitAutoVarCleanups(const AutoVarEmission &emission) {
 
-  const VarDecl &D = *emission.Variable;
+  // If we're initializing directly on the heap, _Block_object_destroy will
+  // handle destruction, so we don't need to perform cleanups directly.
```
I think the right way to think of this is that the stack copy of the byref is always uninitialized in this situation, so there's nothing for us to destroy.  Basically, we make an uninitialized byref structure on the stack, ask the runtime to move it to the heap with a trivial move function, and then initialize the object on the heap directly.

ille-apple replied:
Indeed... and looking at this again, I put too much code inside the if statement.

Not only that, there's an existing bug here when combining `__attribute__((cleanup))` with `__block`: if the variable was copied to the heap, the cleanup function will be called after the byref allocation has already been freed.

Will fix.

rjmccall commented on clang/lib/CodeGen/CGBlocks.cpp:
```diff
@@ -2653,11 +2655,39 @@ const BlockByrefInfo &CodeGenFunction::getBlockByrefInfo(const VarDecl *D) {
   return pair.first->second;
 }
 
+void CodeGenFunction::emitByrefInitOnHeap(llvm::Value *P) {
```
This isn't actually doing the initialization, so I think this is misnamed.  This is just setting up for the initialization by allocating a byref on the heap.

ille-apple replied:
Good point.  I'll change the name to `emitByrefHeapAlloc`.

rjmccall commented on clang/lib/CodeGen/CGBlocks.cpp:
```diff
@@ -2764,6 +2794,9 @@ void CodeGenFunction::emitByrefStructureInit(const AutoVarEmission &emission) {
   }
+
+  if (emission.NeedsInitOnHeap)
+    emitByrefInitOnHeap(pointer);
```
So, this is not exception-safe, right?  If the initialization throws, we have no way to release the byref object we're allocating here.

Now, it looks like the [blocks runtime](https://github.com/llvm/llvm-project/blob/0a93e9f2e264472f1c8c3b5d8726aa43a02ecace/compiler-rt/lib/BlocksRuntime/runtime.c#L368) is itself not exception-safe, so maybe we can hand-wave this?  `_Block_byref_assign_copy ` allocates, calls the copy operation, and makes no effort to deallocate on a throw.  On the other hand, the byref "copy" operation is typically a call to a move constructor, so it is a lot less likely in practice that it'll throw than that the original initialization will.

The only way I can think of to handle this would be to actually put a "successfully initialized" flag in the byref payload and then no-op the destructor if it isn't set.  That's implementable but pretty awful.  So maybe we should just acknowledge it's not exception-safe and suggest that the blocks runtime add new APIs to support this properly.  What do you think?

ille-apple replied:
Argh, you're right.

In principle the `BLOCK_HAS_COPY_DISPOSE` flag could be used for that purpose, instead of creating a separate field in the payload.  It could start out unset (making the runtime skip the call to `byref_dispose`), and then be set after initialization.

But the initialization might send the block capturing the object to another thread.  If initialization then throws, I assume that nobody will subsequently _call_ the block, or if they do, the block won't actually try to use the object.  That's straightforward UB.  But it doesn't seem like it should be UB just to send the block.  So there might be concurrent calls to `_Block_object_dispose`.  Not only that, 'our' `_Block_object_dispose` call (the one that would hypothetically be added on the cleanup path to fix the memory leak) might not be the one to drop the last reference.

Therefore, `flags` would have to be modified atomically, particularly because the same word is also used for the reference count.  And even with an atomic operation, you would still have theoretical UB in the blocks runtime when it reads `flags` non-atomically.  Though, not any more UB than it's already incurring thanks to potential concurrent refcount modifications...  But what if an alternate block runtime tries to cache `flags`?  The only alternate runtime I know of is GNUstep's and it looks like [it doesn't cache it](https://github.com/gnustep/libobjc2/blob/master/blocks_runtime.m#L199), but this still feels very dubious.

So then there's the alternative you suggested of just acknowledging that this isn't exception-safe (specifically, that it leaks memory if initialization throws).  But that makes this whole patch feel like an awkward middle ground.  If self-capturing `__block` variables are not going to be supported properly (pending new APIs), then maybe they shouldn't be supported at all and should just be an error (pending new APIs).

But if new APIs are added, teaching Clang how to recognize when the target OS supports them would be quite a lot of work for what is ultimately an edge case.

I'm going to see how ugly it ends up being to add a "successfully initialized" flag.

ille-apple replied:
Updated to add the flag.  It's not all that bad.

In fact, if, as I suggested in my last comment, we want to support the byref lingering after unwinding, then there *must* be some kind of runtime state – to track whether whoever drops the last reference needs to run the destructor.  It would be more natural for the Blocks runtime to track that state rather than relying on compiler-generated code, but the state has to live somewhere.

If, on the other hand, we don't want to support the byref lingering after unwinding, then I could go back to the idea of mutating the existing `flags` field.  In that scenario, the existence of runtime state would be unnecessary, but the implementation would be relatively unobtrusive.

ille-apple replied:
Ping.  Any thoughts?  I know this is a mess :)

If both options are too messy, there's also the option of making these self-captures a hard error, which would not break backwards compatibility since they never worked properly.

ille-apple replied:
Ping.

rjmccall reviewed (commented):
This generally looks okay, but see the comment about exception safety.

ille-apple committed:
Updates:

- Revamp heap-init dispose helper to use a conditional flag.

- Fix __attribute__((cleanup)).

- Revert the changes from the first commit in this PR that involve
  copying ForceNoopCopy to BlockByrefHelpers.  Upon review, this turned
  out to be unnecessary.

ille-apple committed:
Add note about thread safety

ille-apple commented:
Ping.
